ICR RSE: Improvements and Text Simplifications

Robadob opened the pull request:
Manual merge of the appropriate changes from #51 (cc @msarkis-icr), mostly typographic/simplification changes (those in profiling are mostly directly used, whereas optimisation was heavily conflicted) including a new figure what I've reproduced.

**Todo:**
- [ ] New figure probably needs a bit more iteration (cc @JostMigenda for feedback)
- [x] How should https://github.com/ICR-RSE-Group be added to `learners/acknowledgements.md` (cc @msarkis-icr)
    * e.g. "Members of [The Institute of Cancer Research RSE](https://github.com/ICR-RSE-Group) contributed a large number of typographic corrections and simplifications during the alpha phase of the course's development lifecycle."
    * Happy to also name/tag specific people if preferred.
    * Manual merge has stolen all of your names from git history, if you're keen to rewrite history to change that and submit a new PR go for it (but probably wait for Jost to review the new figure).
- [x] Write a proper commit message
- [ ] Fixup all the correction commits.
    
 Closes #51

JostMigenda commented on episodes/optimisation-numpy.md:
```diff
@@ -28,14 +28,14 @@ Packages like NumPy and Pandas work similarly: They have been written in compile
 [NumPy](https://numpy.org/) is a commonly used package for scientific computing, which provides a wide variety of methods.
 
-It adds restriction via its own [basic numeric types](https://numpy.org/doc/stable/user/basics.types.html) and static arrays to enable even greater performance than that of core Python. However if these restrictions are ignored, the performance can become significantly worse.
+It adds restriction via it's own [basic numeric types](https://numpy.org/doc/stable/user/basics.types.html) and static arrays to enable even greater performance than that of core Python. However if these restrictions are ignored, the performance can become significantly worse.
```
```suggestion
It adds restriction via its own [basic numeric types](https://numpy.org/doc/stable/user/basics.types.html) and static arrays to enable even greater performance than that of core Python. However if these restrictions are ignored, the performance can become significantly worse.
```

JostMigenda commented on episodes/optimisation-numpy.md:
```diff
@@ -379,7 +380,7 @@ Pandas' methods by default operate on columns. Each column or series can be thou
 Following the theme of this episode, iterating over the rows of a data frame using a `for` loop is not advised. The pythonic iteration will be slower than other approaches.
 
-Pandas allows its own methods to be applied to rows in many cases by passing `axis=1`, where available these functions should be preferred over manual loops. Where you can't find a suitable method, `apply()` can be used, which is similar to `map()`, to apply your own function to rows.
+Pandas allows it's own methods to be applied to rows in many cases by passing `axis=1`, where available these functions should be preferred over manual loops. Where you can't find a suitable method, `apply()` can be used, which is similar to `map()`, to apply your own function to rows.
```
```suggestion
Pandas allows its own methods to be applied to rows in many cases by passing `axis=1`, where available these functions should be preferred over manual loops. Where you can't find a suitable method, `apply()` can be used, which is similar to `map()`, to apply your own function to rows.
```

JostMigenda commented on episodes/profiling-introduction.md:
```diff
@@ -41,7 +41,7 @@ Increasingly, particularly with relation to HPC, attention is being paid to the
 Profiling is most relevant to working code, when you have reached a stage that the code works and are considering deploying it.
 
-Any code that will run for more than a few minutes over it's lifetime, that isn't a quick one-shot script can benefit from profiling.
+Any code that will run for more than a few minutes over its lifetime, that isn't a quick one-shot script can benefit from profiling.
```
```suggestion
Any code that will run for more than a few minutes over its lifetime and isn't a quick one-off script can benefit from profiling.
```
Probably a matter of taste (and OED says “one-off” is mainly British English, while “one-shot” is mainly American English?), but this feels more natural to me.

JostMigenda commented on learners/setup.md:
````diff
@@ -22,8 +22,12 @@ This course uses Python and was developed using Python 3.11, therefore it is rec
 
-<!-- conda create -n pando python
-     conda activate pando -->
+If you have `conda` available, you can create and activate a new environment named `py311_env` with Python 3.11 using the following command:
+
+```sh
+conda create --name py311_env python=3.11
+conda activate py311_env
````
```suggestion
If you have `conda` available, you can create and activate a new environment named `p_and_o` with Python 3.13 using the following command:

```sh
conda create --name p_and_o python=3.13
conda activate p_and_o
```
* Use the latest version of Python, like we tell learners to do in a later episode
* More meaningful name for the environment

JostMigenda replied:
> In this case is it not better to specify `python` sans version?

Most of the time, yes; however, there’s always a risk in the first weeks after a major Python release that one of the third-party packages with binary extensions used in the course doesn’t have compatible wheels available yet. For example, Python 3.13 came out in October 2024 and `line_profiler` published the first release with compatible wheels [in early December](https://pypi.org/project/line-profiler/4.2.0/#history).
(It’s possible that the sdist of the previous release would compile without issues; however, I wouldn’t want to rely on that working reliably on all learners’ laptops.)

More generally, I think it’d be good to briefly go over the course after every major Python release, to make sure it’s still up to date. (Especially in the next few releases, where free-threading and the JIT will potentially affect performance characteristics significantly.)

If you prefer, we can keep 3.11 here for now and open a new issue (feel free to assign to me!) to update to 3.13 across the board.

Robadob replied:
I'll setup an issue, would be good to have a documented process for updating Python version e.g. if we're keen for things to be tested, may be possible to semi-automate this with CI.

Robadob commented:
Thanks @JostMigenda for the quick feedback, I'll get this fixed up over the weekend.

Robadob commented on episodes/optimisation-data-structures-algorithms.md:
```diff
@@ -152,20 +155,20 @@ Since Python 3.6, the items within a dictionary will iterate in the order that t
+And if you keep adding more books, at some point you'll move to a larger bookcase with more shelves (and thus more fine-grained sorting), to make sure you don't have too many books on a single shelf.
 
-![A bookshelf corresponding to a Python dictionary.](episodes/fig/bookshelf_dict.jpg){alt="An image of two bookcases, labelled “Sci-Fi” and “Romance”. Each bookcase contains shelves labelled in alphabetical order, with zero or few books on each shelf."}
+![A bookshelf corresponding to a Python dictionary.](episodes/fig/bookshelf_dict.jpg){alt="An image of two bookcases, labelled "Sci-Fi" and "Romance". Each bookcase contains shelves labelled in alphabetical order, with zero or few books on each shelf."}
```
Yeah, I've normally swapped the outer quotes for single quotes in these occasions, clearly messed this one up.

Robadob replied:
Done a test build with single quotes and it outputs correctly.

The double quotes are automatically converted to `&quot;` (probably the real solution for nesting double quotes, I'd just forgot that html symbol existed (did remember `&nbsp;` above))

Robadob commented on episodes/optimisation-introduction.md:
```diff
@@ -27,28 +27,29 @@ We'll talk briefly about some of these external bottlenecks at the end. For now,
+This classic quote among computer scientists emphasises the importance of considering both performance and maintainability when optimising code and prioritising your optimisations.
 
-Advanced optimisations, mostly outside the scope of this course, can increase the cost of maintenance by obfuscating what code is doing. Even if you are a solo-developer working on private code, your future self should be able to easily comprehend your implementation.
+While advanced optimisations may boost performance, they often come at the cost of making the code harder to understand and maintain. Even if you're working alone on private code, your future self should be able to easily understand the implementation. Hence, when optimising, always weigh the potential impact on both performance and maintainability.
```
> Do we actually have any advanced optimisations in this course that hurt maintainability to improve performance?

It's a good point, the only one that comes to mind is the edge-case pandas data-table unwrapping to a dict of dict.

JostMigenda replied:
Yep, a lot of the time, the more performant example code is also cleaner. Can we re-add the “outside the scope of this course”?

Robadob replied:
Difficult to add that back into the new paragraph. So I've added the sentence

>  Such advanced optimisations are mostly outside the scope of this course, but you may already know some.

I think this better covers why it's worth mentioning even if we don't teach. (although "know some" doesn't feel too eloquent).

Robadob commented on episodes/fig/list-append.png:
Do you think it would be worth doing something like $P_1$, $P_2$, $P_3$ to be even clearer that they're pointers (without needing to make the figure massive to draw the data storage?)

JostMigenda replied:
I think that might be clearer, yes; let’s try it.

Also, a minor regression in the updated figure: In the red/purple text (“1 spare element”, “2 new elements”, etc.), “spare” and “new” should start with a lower case letter.

Robadob commented:
I've addressed these points, a few responses to clarify changes though.

Thanks again for catching out all my terrible grammar (illustrator spell-check suggested I capitalise Spare/New [in red/purple] as start of sentence following the numbers).

msarkis-icr commented:
> * [ ]  How should https://github.com/ICR-RSE-Group be added to `learners/acknowledgements.md` (cc @msarkis-icr)
>   
>   * e.g. "Members of [The Institute of Cancer Research RSE](https://github.com/ICR-RSE-Group) contributed a large number of typographic corrections and simplifications during the alpha phase of the course's development lifecycle."
>   * Happy to also name/tag specific people if preferred.
>   * Manual merge has stolen all of your names from git history, if you're keen to rewrite history to change that and submit a new PR go for it (but probably wait for Jost to review the new figure).

Hello @Robadob,
We are happy to name both Anastasiia shcherbakova and myself, Mira Sarkis as contributors from ICR-RSE.

JostMigenda commented on episodes/optimisation-data-structures-algorithms.md:
```diff
@@ -59,7 +59,7 @@ If it doesn't, it will re-allocate a larger array, copy across the elements, and
 
 <!-- Based on ICR-RSE's visual note: https://icr-rse-group.github.io/carpentry-pando-python/optimisation-data-structures-algorithms.html#lists -->
-![A visual diagram of list storage.](episodes/fig/list-append.png){alt="1. A list uses an array for storage, a continuous block of memory. (An array with 5 elements) 2. It can have additional storage, beyond it's length. This makes appending fast. (an array with 7 elements, 2 unoccupied elements are marked '?', '7' is then appended leaving 1 unoccupied element marked '?' in a second array) 3. App ...
+![A visual diagram of list storage.](episodes/fig/list-append.png){alt="1. A list uses an array for storage, a continuous block of memory. (An array with 5 elements) 2. It can have additional storage, beyond its length. This makes appending fast. (an array with 7 elements, 2 unoccupied elements are marked '?', '7' is then appended leaving 1 unoccupied element marked '?' in a second array) 3. Appe ...
```
```suggestion
![A visual diagram of list storage.](episodes/fig/list-append.png){alt="1. A list uses an array for storage, a continuous block of memory. (An array with 5 elements) 2. It can have additional storage, beyond its length. This makes appending fast. (an array with 7 elements, 2 unoccupied elements are marked '?', '6' is then appended leaving 1 unoccupied element marked '?' in a second array) 3. Appending to a full list causes it to grow. This makes some appends slower. (An array with 7 occupied elements, '8' is appended, the second array has 9 elements with the final element marked '?') A list will typically grow by 12.5%, hence shorter lists will grow more frequently when appending." }
```
Also, we should probably change e.g. `6` to `pointer #6` if that works well in the updated figure.

Robadob replied:
I've merged this suggestion, but the alt text feels very unweildy.

Would it better to just describe what the figure demonstrates verbaly? (e.g. this was generated by AI)

> A list uses a contiguous block of memory, similar to an array, for storing the pointers to its elements. It is depicted as a series of five adjacent boxes, each containing 'P' with a subscript from 1 to 5, representing pointers to the list's elements.
> 
> It can have additional storage, beyond its length. This makes appending fast. An illustration shows the previous list with two extra empty boxes marked with question marks, indicating spare elements. Below, Python code `len(my_list) == 5` and `my_list.append(6)` is shown. An arrow points to the list after appending, now showing 'P6' in the first of the previously empty boxes, and one remaining empty box. The length is now `len(my_list) == 6`.
> 
> Appending to a full list causes it to grow. This makes some appends slower. An illustration depicts a full list with 'P1' through 'P7' in adjacent boxes and a label &quot;No spare elements!&quot;. Below, Python code `len(my_list) == 7` and `my_list.append(8)` is shown. An arrow indicates the result: a new, larger contiguous block of memory with 'P1' through 'P8' followed by a question mark in an additional box, indicating one spare element. The label &quot;2 new elements&quot; with curved arrows suggests that when the list grows, it typically allocates more memory than just the space for the new element.
> 
> A concluding note states that a list will typically grow by 12.5%, hence shorter lists will grow more frequently when appending.

JostMigenda replied:
While longer, I find that easier to understand, yes. Just made a few tweaks of my own to shorten it:

> A list uses a contiguous block of memory, similar to an array, for storing the pointers to its elements. It is depicted as a series of five adjacent boxes, labelled 'P1' to 'P5', representing pointers to the list's elements.
>
> It can have additional storage beyond its length to make appends faster. An illustration shows the previous list with two extra empty boxes marked with question marks, indicating spare elements. Below, Python code `len(my_list) == 5` and `my_list.append(6)` is shown. After appending, the first of the previously empty boxes contains 'P6', and the last one remains empty. The length is now `len(my_list) == 6`.
>
> Appending to a full list causes it to grow. This makes some appends slower. An illustration depicts a full list with 'P1' through 'P7' in adjacent boxes and a label "No spare elements!". Below, Python code `len(my_list) == 7` and `my_list.append(8)` is shown. The result is a new, larger continuous block of memory with 'P1' through 'P8' followed by a question mark in an additional box, indicating one spare element. The label "2 new elements" with curved arrows suggests that when the list grows, it typically allocates more memory than just the space for the new element.
>
> A concluding note states that a list will typically grow by 12.5%, hence shorter lists will grow more frequently when appending.

Robadob replied:
I suppose alt would be to just start each paragraph 1), 2) etc.

Robadob commented:
@JostMigenda 

**Changes**
- Figure, Removed the 1 New/Spare caps that I previously added
- Figure, Changed elements from the form $1$ to $P_1$
- Figure alt text, updated text to match elements (not mentioning pointers) 
- Intro, added back a note that advanced optimisations are outside the scope of the course.

**Pending**
- Figure alt-text, use AI suggested alt text it's less unwieldy than what I drafted.
- conda python version, specify 3.13/3.11 or none (which should default latest)

JostMigenda commented on episodes/optimisation-introduction.md:
```diff
@@ -41,7 +41,7 @@ The remaining content is often abstract knowledge, that is transferable to the v
 This classic quote among computer scientists emphasises the importance of considering both performance and maintainability when optimising code and prioritising your optimisations.
 
-While advanced optimisations may boost performance, they often come at the cost of making the code harder to understand and maintain. Even if you're working alone on private code, your future self should be able to easily understand the implementation. Hence, when optimising, always weigh the potential impact on both performance and maintainability.
+While advanced optimisations are may boost performance, they often come at the cost of making the code harder to understand and maintain. Even if you're working alone on private code, your future self should be able to easily understand the implementation. Hence, when optimising, always weigh the potential impact on both performance and maintainability. While this course does not cover most advan ...
```
```suggestion
While advanced optimisations may boost performance, they often come at the cost of making the code harder to understand and maintain. Even if you're working alone on private code, your future self should be able to easily understand the implementation. Hence, when optimising, always weigh the potential impact on both performance and maintainability. While this course does not cover most advanced optimisations, you may already be familiar with and using some.
```

Robadob replied:
🤦 thanks.

Robadob commented:
Thanks, next I'll try and sort out the technical appendix (probably tail end of this week).
